### PR TITLE
migration: Run migration in server environment

### PR DESCRIPTION
**Pull request by efritz**
Run the `migrator` on server startup. Instead of running all processes at once, we start the Postgres process (if one is defined), then run the migrator service as part of synchronous setup.

Once the migrations are complete, the remaining services are added to the goreman process watcher and the application boots as normal. Goreman behavior was altered to support this use case in #28952.

### CODE REVIEW

File: cmd/server/shared/shared.go [hunk 1]
```diff
@@ -194,17 +210,59 @@ func Main() {
...
 		ProcDiedAction: procDiedAction,
-	})
-	if err != nil {
+	}
+	startProcesses(group, "postgres", postgresProcfile, options)
+
+	if runMigrations {
+		// Run migrations before starting up the application but before
+		// starting any postgres instance within the server container.
+		runMigrator()
```
Comment (camdencheek) [inline cmd/server/shared/shared.go:234]: Does `migrator` handle waiting for Postgres to start?

File: cmd/server/shared/shared.go [hunk 2]
```diff
@@ -194,17 +210,59 @@ func Main() {
...
+	options := goreman.Options{
 		RPCAddr:        "127.0.0.1:5005",
 		ProcDiedAction: procDiedAction,
-	})
-	if err != nil {
+	}
+	startProcesses(group, "postgres", postgresProcfile, options)
+
+	if runMigrations {
+		// Run migrations before starting up the application but before
```
Comment (camdencheek) [inline cmd/server/shared/shared.go:232]:
```suggestion
		// Run migrations before starting up the application but after
```

File: cmd/server/shared/shared.go [hunk 3]
```diff
@@ -194,17 +210,59 @@ func Main() {
...
+func runMigrator() {
+	log.Println("Starting migrator")
+
+	var output bytes.Buffer
+	e := execer{Out: &output}
+
+	for _, schemaName := range []string{"frontend", "codeintel"} {
+		e.Command("migrator", "up", "-db", schemaName)
+		if err := e.Error(); err != nil {
+			pgPrintf("Migrating %s schema failed:\n%s", schemaName, output.String())
```
Comment (camdencheek) [inline cmd/server/shared/shared.go:262]: nit: it looks like `pgPrintf` isn't actually intended to handle multi-line format strings. However, it looks like this is happening in multiple places here, so feel free to ignore in this PR.

File: cmd/server/shared/shared.go (hunk 1 above)
Comment (efritz) [reply cmd/server/shared/shared.go:234]: It does the exact same thing as the frontend through connection and _up to_ migration: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/dbconn/open.go?L53

**Review by camdencheek — APPROVED** [review]
Changes look good to my green eyes, but I'd recommend waiting for someone with more context around server to take a peek

File: cmd/server/shared/shared.go [hunk 4]
```diff
@@ -174,16 +177,29 @@ func Main() {
...
 	postgresLine, err := maybePostgresProcFile()
 	if err != nil {
 		log.Fatal(err)
 	}
 	if postgresLine != "" {
-		procfile = append(procfile, postgresLine)
+		if restore {
+			// If in restore mode, only run PostgreSQL
+			procfile = []string{}
+			procfile = append(procfile, postgresLine)
```
Comment (keegancsmith) [inline cmd/server/shared/shared.go:195]: super minor
```suggestion
			procfile = []string{postgresLine}
```

File: cmd/server/shared/shared.go [hunk 5]
```diff
@@ -194,17 +210,59 @@ func Main() {
...
+}
+
+func run(procfile, postgresProcfile []string, runMigrations bool, procDiedAction goreman.ProcDiedAction) {
+	if !runMigrations {
+		procfile = append(procfile, postgresProcfile...)
+		postgresProcfile = nil
 	}
 
-	err = goreman.Start([]byte(strings.Join(procfile, "\n")), goreman.Options{
+	group, _ := errgroup.WithContext(context.Background())
```
Comment (keegancsmith) [inline cmd/server/shared/shared.go:223]: just checking, normally you need the ctx to inform the different goroutines to stop running. But in this case goreman stores the list of subprocesses in global state so if one dies, all are told to shutdown?

File: cmd/server/shared/shared.go (hunk 3 above)
Comment (keegancsmith) [reply cmd/server/shared/shared.go:262]: output will include all output so far, not just for the last `e.Command`

File: cmd/server/shared/shared.go [hunk 6]
```diff
@@ -194,17 +210,59 @@ func Main() {
...
+
+	log.Printf("Starting %s processes\n", name)
+	group.Go(func() error { return goreman.Start([]byte(strings.Join(procfile, "\n")), options) })
+}
+
+func runMigrator() {
+	log.Println("Starting migrator")
+
+	var output bytes.Buffer
+	e := execer{Out: &output}
```
Comment (keegancsmith) [inline cmd/server/shared/shared.go:257]: by setting `output` we won't get to stderr/stdout as these commands run. For long running migrations, I assume the output from migrator would be useful if someone is watching what is going on?

File: cmd/server/shared/shared.go [hunk 7]
```diff
@@ -194,17 +210,59 @@ func Main() {
...
 		log.Fatal(err)
 	}
 }
+
+func startProcesses(group *errgroup.Group, name string, procfile []string, options goreman.Options) {
+	if len(procfile) == 0 {
+		return
+	}
+
+	log.Printf("Starting %s processes\n", name)
```
Comment (keegancsmith) [inline cmd/server/shared/shared.go:249]: nit: you don't need newlines for log pkg
```suggestion
	log.Printf("Starting %s processes", name)
```

File: cmd/server/shared/shared.go [hunk 8]
```diff
@@ -194,17 +210,59 @@ func Main() {
...
+	group.Go(func() error { return goreman.Start([]byte(strings.Join(procfile, "\n")), options) })
```
Comment (keegancsmith) [inline cmd/server/shared/shared.go:250]: in normal mode, if postgres fails to start straight away are there any weird race conditions when we then start up `all` after that? IE imagine this order of events

1. postgres start
2. postgres fail
3. all start

We would want all to shutdown nearly straight away. But I imagine we may be in the case where all starts up and everything just fails in a weird way after some time due to postgres being down.

**Review by keegancsmith — APPROVED** [review]

File: cmd/server/shared/shared.go (hunk 5 above)
Comment (efritz) [reply cmd/server/shared/shared.go:222]: Yes (unless you ignore process death via environment variable), and there wasn't a way to feed a context into goreman without more refactors.

File: cmd/server/shared/shared.go (hunk 6 above)
Comment (efritz) [reply cmd/server/shared/shared.go:256]: Oh I wasn't aware, will update that.

File: cmd/server/shared/shared.go (hunk 8 above)
Comment (efritz) [reply cmd/server/shared/shared.go:249]: (Unless process death is ignored via environment variable...) ...the first process to die will take everything down.

File: cmd/server/shared/shared.go (hunk 3 above)
Comment (efritz) [reply cmd/server/shared/shared.go:261]: See https://github.com/sourcegraph/sourcegraph/pull/28929#discussion_r768313623.

**Pull request merged**